airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java

```diff
@@ -123,103 +131,98 @@ protected void startTracked() throws Exception {
-      LOGGER.info("Buffer creation for stream {}.", pair);
-      try {
-        final OnDiskQueue writeBuffer = new OnDiskQueue(queueRoot.resolve(pair.getName()), pair.getName());
-        pairToWriteBuffer.put(pair, writeBuffer);
-      } catch (Exception e) {
-        LOGGER.error("Error creating buffer: ", e);
-      }
-    }
-    LOGGER.info("Buffer creation completed.");
-
     onStart.call();
-    LOGGER.info("write buffers: {}", pairToWriteBuffer.keySet());
-    writerPool.scheduleWithFixedDelay(
-        () -> writeStreamsWithNRecords(MIN_RECORDS, pairToWriteBuffer, recordWriter),
-        THREAD_DELAY_MILLIS,
-        THREAD_DELAY_MILLIS,
-        TimeUnit.MILLISECONDS);
   }
 
   @Override
   protected void acceptTracked(AirbyteMessage message) throws Exception {
     Preconditions.checkState(hasStarted, "Cannot accept records until consumer has started");
 
-    if (message.getType() != AirbyteMessage.Type.RECORD) {
-      return;
+    if (message.getType() == Type.RECORD) {
+      final AirbyteRecordMessage recordMessage = message.getRecord();
+      final AirbyteStreamNameNamespacePair stream = AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage);
+      final String data = Jsons.serialize(message);
+
+      if (!pairs.contains(stream)) {
+        throwUnrecognizedStream(catalog, message);
+      }
+
+      if (!isValidRecord.apply(data)) {
+        pairToIgnoredRecordCount.put(stream, pairToIgnoredRecordCount.getOrDefault(stream, 0L) + 1L);
+        return;
+      }
+
+      if (queue.size() == BATCH_SIZE) {
+        flushQueueToDestination();
+      }
+
+      if (!queue.offer(message)) {
+        throw new IllegalStateException("Could not accept record despite waiting.");
+      }
+    } else if (message.getType() == Type.STATE) {
+      pendingState = message.getState();
+    } else {
+      LOGGER.warn("Unexpected message: " + message.getType());
     }
 
-    // ignore other message types.
-    final AirbyteStreamNameNamespacePair pair = AirbyteStreamNameNamespacePair.fromRecordMessage(message.getRecord());
-    if (!pairs.contains(pair)) {
-      throw new IllegalArgumentException(
-          String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
-              Jsons.serialize(catalog), Jsons.serialize(message)));
+  }
+
+  private void flushQueueToDestination() throws Exception {
+    final List<AirbyteMessage> queueContents = new ArrayList<>();
+    queue.drainTo(queueContents);
+    final Map<Type, List<AirbyteMessage>> recordsByType = queueContents
+        .stream()
+        .collect(Collectors.groupingBy(AirbyteMessage::getType));
+
+    final List<AirbyteRecordMessage> records = recordsByType.getOrDefault(Type.RECORD, new ArrayList<>())
+        .stream()
+        .map(AirbyteMessage::getRecord)
+        .collect(Collectors.toList());
+
+    final Map<AirbyteStreamNameNamespacePair, List<AirbyteRecordMessage>> recordsByStream = records.stream()
```

Comment:
Instead of doing all this complex logic, why do just partition your queue?

Reply:
Are you talking about the record type stuff (that part i 100% agree) or partitioning by stream? In the latter case, I don't think that gets us anything (or really matters). Marginally easier to check the queue size with it not partitioned.